boggle_solver.js

```diff
@@ -38,70 +38,71 @@ function insert(root, word) {
     curr.is_word = true;
 }
 
-function safe(i, j, visited) {
+function safe(i, j, visited, M) {
     return (i >= 0
         && i < M && j >= 0
-        && j < N
+        && j < M
         && !visited[i][j]);
 }
 
 function search(curr, boggle, i, j, visited, str, res)
 {
         if (curr.is_word === true && !(res.includes(str)) && str.length > 2)
         res.push(str)
+    
+    let M = boggle.length
 
-
-    if (safe(i, j, visited))
+    if (safe(i, j, visited, M))
         visited[i][j] = true
 
 
     for (let child in curr.children)
         if (curr.children[child] != null) {
-            ch = child
+            let ch = child
 
-            if (safe(i + 1, j + 1, visited)
+            if (safe(i + 1, j + 1, visited, M)
```

Comment:
It would save more space to loop through a list of neighbors than have if statements for each neighbor